># 이미지 식별하기

- 기존 분석과 마찬가지로, 이미지의 화소도 숫자의 나열로 변환 가능
- 8X8 화소의 이미지라면 일렬로 나열했을 때 1X64, 즉 64개의 화소와 특징량을 지닌 벡터로 정의 가능
- 이는 63차원의 식별 경계를 만들수 있다는 의미를 지님 (2차원은 직선의 식별경계, 3차원은 평면 식별경계, n차원은 n-1 식별경계...)
- 과거 손글씨 데이터로 식별을 수행하면 다음과 같음

In [5]:
from sklearn.datasets import load_digits
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import MinMaxScaler

# 데이터 세트 호출
X_dataset, y_dataset = load_digits(return_X_y=True)

# 데이터 분할
X_train, X_test, y_train, y_test = train_test_split(X_dataset, y_dataset,
                                                   test_size= 0.2, random_state = 42)
print(X_train[0])

# 정규화
scaler = MinMaxScaler()
X_train = scaler.fit_transform(X_train)
X_test = scaler.fit_transform(X_test)

print('='*20)
print(X_train[0])

[ 0.  0.  3. 14.  1.  0.  0.  0.  0.  0. 12. 12.  0.  0.  0.  0.  0.  3.
 16.  6.  0.  0.  0.  0.  0.  5. 16.  2.  0.  0.  0.  0.  0.  6. 16.  2.
  5.  2.  0.  0.  0.  4. 16.  2. 12. 15.  2.  0.  0.  1. 14. 13.  2. 13.
 11.  0.  0.  0.  3. 11. 16. 13.  4.  0.]
[0.         0.         0.1875     0.875      0.0625     0.
 0.         0.         0.         0.         0.75       0.75
 0.         0.         0.         0.         0.         0.1875
 1.         0.375      0.         0.         0.         0.
 0.         0.33333333 1.         0.125      0.         0.
 0.         0.         0.         0.42857143 1.         0.125
 0.3125     0.125      0.         0.         0.         0.25
 1.         0.125      0.75       0.9375     0.125      0.
 0.         0.0625     0.875      0.8125     0.125      0.8125
 0.6875     0.         0.         0.         0.1875     0.6875
 1.         0.8125     0.25       0.        ]


- 학습 모델은 퍼셉트론을 사용
- 퍼셉트론은 `sklearn.linear_model` 모듈의 `Perceptron`을 호술하여 식별 가능

In [6]:
from sklearn.linear_model import Perceptron

classifier = Perceptron(random_state = 42)

classifier.fit(X_train, y_train)
classifier.score(X_test, y_test)

/usr/local/envs/py3env/lib/python3.5/site-packages/sklearn/linear_model/stochastic_gradient.py:128: FutureWarning: max_iter and tol parameters have been added in <class 'sklearn.linear_model.perceptron.Perceptron'> in 0.19. If both are left unset, they default to max_iter=5 and tol=None. If tol is not None, max_iter defaults to max_iter=1000. From 0.21, default max_iter will be 1000, and default tol will be 1e-3.
  "and default tol will be 1e-3." % type(self), FutureWarning)


0.9305555555555556

- 93% 정답률은 언듯 높아보이지만, 목적에 따라 사용 불가능할 수도 있음
- 예를 들어, 93% 정답률로는 11개의 숫자를 지닌 전화번호 중 하나는 틀릴 가능성이 높으므로, 전화번호 식별에 사용하기에는 부적절함